autodiff/reverse/var/var.hpp

```diff
@@ -911,19 +1058,168 @@ struct Hypot3Expr : TernaryExpr<T>
 
     Hypot3Expr(const T& v, const ExprPtr<T>& ll, const ExprPtr<T>& cc, const ExprPtr<T>& rr) : TernaryExpr<T>(v, ll, cc, rr) {}
 
-    virtual void propagate(const T& wprime)
+    void propagate(const T& wprime) override
     {
         l->propagate(wprime * l->val / val);
         c->propagate(wprime * c->val / val);
         r->propagate(wprime * r->val / val);
     }
 
-    virtual void propagatex(const ExprPtr<T>& wprime)
+    void propagatex(const ExprPtr<T>& wprime) override
     {
         l->propagatex(wprime * l / val);
         c->propagatex(wprime * c / val);
         r->propagatex(wprime * r / val);
     }
+
+    void update() override
+    {
+        l->update();
+        c->update();
+        r->update();
+        this->val = hypot(l->val, c->val, r->val);
+    }
+};
+
+// Any expression yielding a boolean depending on arithmetic subexpressions
+struct BooleanExpr
+{
+    template<typename Comparator>
+    struct Negate {
+      template<typename T>
+      bool operator () (const T& a, const T& b) const {
+        return !(Comparator {}(a, b));
+      }
+    };
+
+    bool val = {};
+
+    explicit BooleanExpr(bool v) : val(v) {}
+    operator bool() const { return val; }
+
+    virtual void update() = 0;
+    virtual std::shared_ptr<BooleanExpr> wrap() const = 0;
```

Comment:
What's the `wrap` method for?

Reply:
The `wrap` function is a shorthand for generating a shared pointer to base from a derived object instance. I noticed that this pattern came up a lot in the `Expr<T>` free functions and wanted to factor it out. It could be better in principle, a CRTP base could supply it instead, if you prefer.

Reply:
Ah, OK, usually such kind of method is called `clone` instead of `wrap` (at least from my experience). No need for CRTP here (I also avoided it for the forward types `dual` and `real` -- the implementation there relies heavily on SFINAE and `if constexpr` though, but I find it way more intuitive. Looking forward to concepts in C++20!